File: strategy/generator/signal/neutrality/tii_neutrality_cross.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter
from core.models.smooth import Smooth
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class TiiNeutralityCrossSignal(Signal):
    type: SignalType = SignalType.TiiNeutralityCross
    smooth_type: Parameter = StaticParameter(Smooth.SMA)
    major_period: Parameter = StaticParameter(40.0)
    minor_period: Parameter = StaticParameter(20.0)

File: ta_lib/strategies/signal/src/neutrality/tii_neutrality_cross.rs
use base::prelude::*;
use core::prelude::*;
use momentum::tii;

pub struct TiiNeutralityCrossSignal {
    smooth_type: Smooth,
    major_period: usize,
    minor_period: usize,
}

impl TiiNeutralityCrossSignal {
    pub fn new(smooth_type: Smooth, major_period: f32, minor_period: f32) -> Self {
        Self {
            smooth_type,
            major_period: major_period as usize,
            minor_period: minor_period as usize,
        }
    }
}

impl Signal for TiiNeutralityCrossSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.minor_period, self.major_period)
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let tii = tii(
            &data.close(),
            self.smooth_type,
            self.major_period,
            self.minor_period,
        );

        (
            tii.cross_over(&NEUTRALITY_LINE),
            tii.cross_under(&NEUTRALITY_LINE),
        )
    }
}
